### PR TITLE
HPCC-23526 Roxie should resolve topology name more often

### PR DESCRIPTION
**Pull request by richardkchapman**
In k8s settings, the topology server name may refer to a different IP if
topology server restarts, and the dns entry may not have been set up at the
time roxie is starting. Delay resolution of the name until when we are
actually making the call.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [x] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
Tested manually while developing/testing helm templates

### CODE REVIEW

**Comment by richardkchapman** [comment]
@mckellyln Please review

**Comment by richardkchapman** [comment]
I have not yet tested the headless server support part of this PR

File: roxie/udplib/udptopo.cpp [hunk 1]
```diff
@@ -225,70 +226,78 @@ class TopologyManager
...
-                else
-                {
-                    if (responseLen > maxReasonableResponse)
+                    unsigned topoBufLen = md5.length()+topoBuf.length();
+                    _WINREV(topoBufLen);
+                    topo->write(&topoBufLen, 4);
+                    topo->write(md5.str(), md5.length());
+                    topo->write(topoBuf.str(), topoBuf.length());
+                    unsigned responseLen;
+                    topo->read(&responseLen, 4);
```
Comment (mckellyln) [inline roxie/udplib/udptopo.cpp]: Perhaps should use readtms() here.

File: roxie/udplib/udptopo.cpp [hunk 2]
```diff
@@ -225,70 +226,78 @@ class TopologyManager
...
-                                    Owned<const ITopologyServer> newServer = new CTopologyServer(eol);
-                                    SpinBlock b(lock);
-                                    currentTopology.swap(newServer);
+                                    const char *eol = strchr(mem, '\n');
+                                    if (eol)
+                                    {
+                                        eol++;
+                                        md5.clear().append(eol-mem, mem);  // Note: includes '\n'
+                                        Owned<const ITopologyServer> newServer = new CTopologyServer(eol);
+                                        SpinBlock b(lock);
```
Comment (mckellyln) [inline roxie/udplib/udptopo.cpp]: We should perhaps limit use of spinblock in the future.

Comment (richardkchapman) [reply roxie/udplib/udptopo.cpp]: Agree. This is not new code though

File: system/jlib/jsocket.cpp [hunk 3]
```diff
@@ -6301,6 +6301,87 @@ inline bool appendv4range(SocketEndpointArray *array,char *str,SocketEndpoint &e
...
+bool SocketEndpointArray::fromName(const char *name, unsigned defport)
+{
+    // Lookup a single name that may resolve to multiple IPs in a headless service scenario
+    StringArray portSplit;
+    portSplit.appendList(name, ":");
+    switch (portSplit.ordinality())
+    {
+    case 2:
+        defport = atoi(portSplit.item(1));
+        name = portSplit.item(0);
```
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: Are we changing name here for caller ?

Comment (richardkchapman) [reply system/jlib/jsocket.cpp]: No, just using the bit before : in the rest of this function

**Review by mckellyln — APPROVED** [review]

**Comment by mckellyln** [comment]
@richardkchapman looks good.

**Comment by AttilaVamos** [comment]
Strange, because the disk isn't full.

**Pull request merged**